### DISCUSSION TITLE
Comprehensive unit testing

### DISCUSSION THREAD

**Discussion by mzabaluev** [movementlabsxyz/movement#394] [Ideas]
We need to improve protocol unit tests and get them run in CI.

Carrying over notes from #393:

Unit tests are written to support development and check for regressions, mainly in protocol units and utility crates.

* No goal to reach comprehensive coverage, but essential use cases need to be tested.
* Regression testing normally goes here.
* Property testing is more practical on level of individual crates.

## No external infrastructure

Currently, our tests include infrastructure-dependent integration style tests, requiring local or remote service setup to run and feature-gated out of the default `cargo test` run.  This should be avoided and mocking should be implemented instead.

One exception is boostrapping of file storage, configuration files, and similar as needed for test execution. This should be performed in temporary directories that are cleaned up at the end of a test.

More heavyweight local setup (e.g. bringing up Anvil and deploying the MCR contract) is possible, but it's impractical to do for every test. Reusing the infra between test cases is possible as a once-init, but this creates test dependency issues.

# Short-term priorities

* Try to reduce infrastructure-dependent tests, replacing the dependencies with mocks or promoting such tests to full e2e (discussed in #395).
* Run all available unit tests and "light integration" tests regularly in CI.
* Have at least some tests run in CI with coverage metrics.

**Comment by l-monninger** [comment]
So, I think we can start identifying tasks here. Can you list areas you would want to prioritize? Wd can then use learnings from improving tests there to improve elsewhere.

**Reply by mzabaluev** [reply]
I have updated the short-term priority list. Please check if it's OK or if something need to be added (long-shot tasks?).
My current opinion is that we should not have many infrastructure-dependent tests, but if it's easier (e.g. to use Eth client against Anvil rather than simulate a particular response), we'll need another category and tasks for maintaning such tests in CI.
